raiden/tests/utils/factories.py

```diff
@@ -28,20 +29,59 @@
     NettingChannelEndState,
     NettingChannelState,
     RouteState,
+    TokenNetworkState,
     TransactionExecutionStatus,
     message_identifier_from_prng,
 )
+from raiden.transfer.state_change import ContractReceiveChannelNew, ContractReceiveRouteNew
 from raiden.transfer.utils import hash_balance_data
 from raiden.utils import CanonicalIdentifier, privatekey_to_address, random_secret, sha3, typing
 from raiden.utils.signer import LocalSigner, Signer
 
-EMPTY = object()
+EMPTY = 'empty'
+GENERATE = 'generate'
+
+
+def _partial_dict(full_dict: Dict, *args) -> Dict:
+    return {key: full_dict[key] for key in args}
+
+
+class Properties:
+    """ Base class for all properties classes. """
+    DEFAULTS: ClassVar['Properties'] = None
+    TARGET_TYPE: ClassVar[Type] = None
+
+    @property
+    def kwargs(self):
+        return {key: value for key, value in self.__dict__.items() if value is not EMPTY}
+
+    def extract(self, subset_type: Type) -> 'Properties':
+        field_names = [field.name for field in fields(subset_type)]
+        return subset_type(**_partial_dict(self.__dict__, *field_names))
+
+    def partial_dict(self, *args) -> Dict[str, Any]:
+        return _partial_dict(self.__dict__, *args)
 
 
 def if_empty(value, default):
     return value if value is not EMPTY else default
 
 
+def _replace_properties(properties, defaults):
+    replacements = {
+        k: create_properties(v, defaults.__dict__[k]) if isinstance(v, Properties) else v
+        for k, v in properties.kwargs.items()
+    }
+    return replace(defaults, **replacements)
+
+
+def create_properties(properties: Properties, defaults: Properties = None) -> Properties:
+    full_defaults = type(properties).DEFAULTS
```

Comment:
why are we using `type()` here? `properties.DEFAULTS` should use the class values if the instance does not overwrite it.